### PR TITLE
feat: Add Rake instrumentation

### PR DESCRIPTION
**Pull request by yoheyk**
This PR implements #17 

Porting the [datadog instrumentation](https://github.com/DataDog/dd-trace-rb/blob/v1.1.0/lib/datadog/tracing/contrib/rake/instrumentation.rb) (Arguments have to be quantized so ignoring them per the [porting guide](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/83adb0edd2c9c648f7df64e4952317ee39b14dec/adapters/datadog-porting-guide.md#things-to-ignore-for-now))

There are two methods to invoke rake tasks, `invoke` and `execute` but the `invoke` method is calling the `execute` method inside:
https://github.com/ruby/rake/blob/v13.0.6/lib/rake/task.rb#L219

I couldn't find a suitable one in [Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) and [this discussion](https://github.com/open-telemetry/opentelemetry-specification/pull/1582) doesn't seem to reach a conclusion yet so I am setting the following span kinds for now (but any suggestions are welcome!):
- `invoke` : `internal` span kind
- `execute`: `client` span kind

Here is a screenshot of a local trace:
localhost_16686_trace_408c3dadfc2630dd1c41959f9223655d

### CODE REVIEW

**Comment by ericmustin** [comment]
Hey thanks for the PR! just ack'ing this and assigning it to myself for review, will try to take a look shortly. Appreciate you including a screenshot of a trace UI :)

**Review by ericmustin — COMMENTED** [review]
Hey! this broadly looks great. Appreciate the contribution. I think this is super close! Left one or two small nits / questions but nothing major

re:
> I couldn't find a suitable one in [Semantic Conventions](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace/semantic_conventions) and https://github.com/open-telemetry/opentelemetry-specification/pull/1582 doesn't seem to reach a conclusion yet so I am setting the following span kinds for now (but any suggestions are welcome!):

> invoke : internal span kind
> execute: client span kind

I will have to consult the spec here a bit more closely to try to understand what the appropriate span kind should be. my instinct says this should follow messaging convention semantics but I agree that it's not exactly clear/a perfect fit. I will try to get a clear answer on this (or see what some other languages with make-like program instrumentions have done) and follow up.

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb
Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb:14]: Confirmed this is compatible with the version `0.9.0` and above so added the check: 20957da69b0da7e4065411c340abb5c007a1a1c6

> I see in the gemspec there's a dev dependency on 13.x but that seems a bit restrictive, is there an intention here?

Not really, I just set the latest major version 😄. Let me know if I should change it.

**Review by ericmustin — APPROVED** [review]
Thanks for addressing all my feedback LGTM!

It would be great if either @arielvalentin or @plantfansam could give this a review as well as i'm still ramping back up wrt the `-contrib` repo.

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb
Comment (arielvalentin) [inline instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb:11]: Since this is only one key-value pair how do you feel about inlining this invocation?

```suggestion
            tracer.in_span('rake.invoke', attributes: { 'rake.task' => name }) do
```

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb
Comment (arielvalentin) [inline instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb:18]: ♻️ 
```suggestion
            tracer.in_span('rake.execute', attributes: { 'rake.task' => name }) do
```

File: instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb
Comment (arielvalentin) [inline instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb:9]: Would a better place to load these be in the test helper by `require 'opentelemetry-instrumentation-rake'` or using `Bundler.require`?

**Review by arielvalentin — CHANGES_REQUESTED** [review]
Thank you for your submission and patience. I was unavailable for reviews for a few weeks and have some feedback for you in this PR. 

Some of the comments are small suggestions that you may want to consider but are not required, however I am concerned about adding an `at_exit`[^1] hook as part of the instrumentation as opposed to making it part of the SDK package and would like some insight from the @open-telemetry/ruby-maintainers 

[^1]: https://blog.arkency.com/2013/06/are-we-abusing-at-exit/

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb
Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb]: updated in 9937238

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb
Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/patches/task.rb]: updated in 9937238

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb
Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb:45]: Thanks for your feedback @arielvalentin !
I'll wait for their response before making any change here.

File: instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb
Comment (yoheyk) [reply instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb:9]: this is from the instrumentation generator:
https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/1af9fc1a35264dcaf3bd0d88234e8ad8dacdaa22/.instrumentation_generator/templates/test/instrumentation.rb.tt#L9

and other instrumentations are doing the same...

but should I modify it?

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb
Comment (ahayworth) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb:45]: A similar situation occurred in the ActiveJob instrumentation (although that was a `force_flush` and not a shutdown).

1. Maybe force_flush is a more palatable option?
2. We decided to hide that behind a configuration flag: https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb#L63

File: instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb
Comment (arielvalentin) [reply instrumentation/rake/test/opentelemetry/instrumentation/rake/instrumentation_test.rb:9]: No worries. I will submit a separate PR to leverage bundler instead.

File: instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb
Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb]: I took another look at the datadog instrumentation and it is calling `shutdown` in the ensure blocks:
https://github.com/DataDog/dd-trace-rb/blob/v1.1.0/lib/datadog/tracing/contrib/rake/instrumentation.rb#L29

I made a change to call `force_flush` in the eusure blocks instead of adding an `at_exit` hook: cba89c36548678c10f2a0da599a87b484996319b

how is this?

Comment (yoheyk) [reply instrumentation/rake/lib/opentelemetry/instrumentation/rake/instrumentation.rb]: @arielvalentin I have updated this PR to not use `at_exit` hook but let me know if there is anything I need to change. Thanks!!

**Comment by jfeltesse-mdsol** [comment]
@arielvalentin is there anything else that needs to be addressed?

**Review by arielvalentin — APPROVED** [review]

**Comment by arielvalentin** [comment]
@yoheyk thank you for your patience. I will merge as soon as the builds are done!

**Comment by yoheyk** [comment]
@arielvalentin Thanks!!

**Comment by amber-h** [comment]
this is very exciting!

**Pull request merged**
